Lib/ntpath.py

```diff
@@ -554,36 +554,48 @@ def normpath(path):
         return prefix + sep.join(comps)
 
 
-def _abspath_fallback(path):
-    """Return the absolute version of a path as a fallback function in case
-    `nt._getfullpathname` is not available or raises OSError. See bpo-31047 for
-    more.
-
-    """
-
-    path = os.fspath(path)
-    if not isabs(path):
-        if isinstance(path, bytes):
-            cwd = os.getcwdb()
-        else:
-            cwd = os.getcwd()
-        path = join(cwd, path)
-    return normpath(path)
-
 # Return an absolute path.
 try:
     from nt import _getfullpathname
 
 except ImportError: # not running on Windows - mock up something sensible
-    abspath = _abspath_fallback
+    def abspath(path):
+        """Return the absolute version of a path."""
+        path = os.fspath(path)
+        if not isabs(path):
+            if isinstance(path, bytes):
+                cwd = os.getcwdb()
+            else:
+                cwd = os.getcwd()
+            path = join(cwd, path)
+        return normpath(path)
 
 else:  # use native Windows method on Windows
     def abspath(path):
         """Return the absolute version of a path."""
         try:
             return _getfullpathname(normpath(path))
         except (OSError, ValueError):
-            return _abspath_fallback(path)
+            # See gh-75230, handle outside for cleaner traceback
+            pass
+        path = os.fspath(path)
+        if not isabs(path):
+            if isinstance(path, bytes):
+                sep = b'\\'
+                cwd = os.getcwdb()
+            else:
+                sep = '\\'
+                cwd = os.getcwd()
+            drive, root, path = splitroot(path)
+            if drive and drive != splitroot(cwd)[0]:
+                try:
+                    path = join(_getfullpathname(drive), path)
+                except (OSError, ValueError):
+                    # Invalid drive \x00: on Windows; assume root directory
+                    path = drive + sep + path
+            else:
+                path = join(cwd, root + path)
+        return normpath(path)
```

Comment:
The C implementations of `splitroot()`, `normpath()`, and `_getfullpathname()` have to decode and then encode a `bytes` path. How about trying to minimize this? If the implementation operates on the path as a Unicode string, it shouldn't be an issue in terms of the wrong path type in an `OSError`. Exceptions from `_getfullpathname()` are handled, and `normpath()` and `splitroot()` shouldn't raise `OSError`. 

For example:

```python
    def abspath(path):
        """Return the absolute version of a path."""
        if is_bytes := isinstance(path := os.fspath(path), bytes):
            path = os.fsdecode(path)
        path = normpath(path)
        try:
            path = _getfullpathname(path)
        except (OSError, ValueError):
            pass
        if not isabs(path):
            drive, root, path = splitroot(path)
            if drive:
                try:
                    path = join(_getfullpathname(drive), path)
                except (OSError, ValueError):
                    path = drive + sep + path
            else:
                path = join(os.getcwd(), root + path)
            path = normpath(path)
        if is_bytes:
            return os.fsencode(path)
        return path
```

Reply:
Hmm, I thought calling `os.fspath()` & `os.fsdecode()` was much slower than letting argument clinic do its thing.
Also, the performance of the fallback function doesn't matter that much, it won't be called frequently. I can do a benchmark if you want though. I'll say that this looks a lot more complicated.

Reply:
It's all good.